BUG/TST: cluster: Occasional failure of a couple tests in `cluster`

I get occasional failures of the test functions `test_kmeans2_high_dim()` and `test_kmeans2_rank1_2()`, which are methods of `scipy.cluster.tests.test_vq.TestKMeans`, defined in `scipy/cluster/tests/test_vq.py`.

The problem is that the call of `kmean2()` generates a warning:

    UserWarning: One of the clusters is empty. Re-run kmeans with a different initialization.

The occurrences are not common; I might have to run rerun the test a dozen times or more between occurrences.

Two examples:

<details>

<summary> $ spin test -m full -j 12 -s cluster </summary>

```
$ spin test -m full -j 12 -s cluster
Invoking `build` prior to running tests:
$ meson compile -j 12 -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/warren/py3.13.7/bin/ninja -C /home/warren/repos/git/forks/scipy-fresh/build -j 12
ninja: Entering directory `/home/warren/repos/git/forks/scipy-fresh/build'
[2/338] Generating subprojects/highs/src/HConfig.h with a custom command
$ meson install --only-changed -C build --destdir ../build-install --tags=runtime,python-runtime,tests,devel
$ export PYTHONPATH="/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages"
$ /home/warren/py3.13.7/bin/python3.13 -P -c 'import scipy'
$ cd /home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages
$ /home/warren/py3.13.7/bin/python3.13 -P -m pytest -n 12 --pyargs scipy.cluster
================================== test session starts ===================================
platform linux -- Python 3.13.7, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/warren/repos/git/forks/scipy-fresh
configfile: pytest.ini
plugins: xdist-3.8.0, anyio-4.11.0, timeout-2.4.0, hypothesis-6.138.14, cov-6.2.1
12 workers [217 items]    
.................................................................................. [ 37%]
.................................................................................. [ 75%]
.................................................F...                              [100%]
======================================== FAILURES ========================================
_________________________ TestKMeans.test_kmeans2_rank1_2[numpy] _________________________
[gw4] linux -- Python 3.13.7 /home/warren/py3.13.7/bin/python3.13

self = <scipy.cluster.tests.test_vq.TestKMeans object at 0x7e855d191950>
xp = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

    def test_kmeans2_rank1_2(self, xp):
        data = xp.asarray(TESTDATA_2D)
        data1 = data[:, 0]
>       kmeans2(data1, 2, iter=1)

data       = array([[-2.2 ,  1.17],
       [-1.63,  1.69],
       [-2.04,  4.38],
       [-3.09,  0.95],
       [-1.7 ,  4.79],
   ...1],
       [-2.88,  1.56],
       [-1.63,  1.46],
       [-1.41,  0.68],
       [-1.96,  2.76],
       [-1.61,  2.11]])
data1      = array([-2.2 , -1.63, -2.04, -3.09, -1.7 , -1.68, -2.26, -2.29, -1.72,
       -1.99, -2.75, -2.45, -4.26, -1.57, -1.96,...-1.29, -2.34, -1.74, -1.97,
       -2.51, -1.64, -2.99, -1.53, -2.24, -2.46, -2.88, -1.63, -1.41,
       -1.96, -1.61])
self       = <scipy.cluster.tests.test_vq.TestKMeans object at 0x7e855d191950>
xp         = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

scipy/cluster/tests/test_vq.py:324: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
scipy/_lib/_util.py:346: in wrapper
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
        NEW_NAME   = 'rng'
        args       = (array([-2.2 , -1.63, -2.04, -3.09, -1.7 , -1.68, -2.26, -2.29, -1.72,
       -1.99, -2.75, -2.45, -4.26, -1.57, -1.96...9, -2.34, -1.74, -1.97,
       -2.51, -1.64, -2.99, -1.53, -2.24, -2.46, -2.88, -1.63, -1.41,
       -1.96, -1.61]), 2)
        as_new_kwarg = False
        as_old_kwarg = False
        as_pos_arg = False
        cmn_msg    = 'To silence this warning and ensure consistent behavior in SciPy None, control the RNG using argument `rng`. Arguments...ndom` or `RandomState` instances will result in an error. See the documentation of `default_rng` for more information.'
        emit_warning = False
        end_version = None
        fun        = <function kmeans2 at 0x7e857fd62660>
        global_seed_set = False
        kwargs     = {'iter': 1}
        old_name   = 'seed'
        position_num = None
scipy/cluster/vq.py:827: in kmeans2
    miss_meth()
        _          = 0
        check_finite = True
        code_book  = array([ 0.47783387, -0.97021566])
        d          = 1
        data       = array([-2.2 , -1.63, -2.04, -3.09, -1.7 , -1.68, -2.26, -2.29, -1.72,
       -1.99, -2.75, -2.45, -4.26, -1.57, -1.96,...-1.29, -2.34, -1.74, -1.97,
       -2.51, -1.64, -2.99, -1.53, -2.24, -2.46, -2.88, -1.63, -1.41,
       -1.96, -1.61])
        has_members = array([False,  True])
        init_meth  = <function _krandinit at 0x7e85887cf1a0>
        iter       = 1
        k          = 2
        label      = array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,..., 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1], dtype=int32)
        minit      = 'random'
        miss_meth  = <function _missing_warn at 0x7e857fd623e0>
        missing    = 'warn'
        nc         = 2
        new_code_book = array([ 0.     , -2.43295])
        rng        = RandomState(MT19937) at 0x7E85CD574E40
        thresh     = 1e-05
        xp         = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _missing_warn():
        """Print a warning when called."""
>       warnings.warn("One of the clusters is empty. "
                      "Re-run kmeans with a different initialization.",
                      stacklevel=3)
E       UserWarning: One of the clusters is empty. Re-run kmeans with a different initialization.


scipy/cluster/vq.py:634: UserWarning
================================ short test summary info =================================
FAILED scipy/cluster/tests/test_vq.py::TestKMeans::test_kmeans2_rank1_2[numpy] - UserWarning: One of the clusters is empty. Re-run kmeans with a different initializat...
============================= 1 failed, 216 passed in 2.37s ==============================
```

</details>

<details>

<summary> $ spin test -m full -j 3 -s cluster </summary>

```
$ spin test -m full -j 3 -s cluster
Invoking `build` prior to running tests:
$ meson compile -j 12 -C build
INFO: autodetecting backend as ninja
INFO: calculating backend command to run: /home/warren/py3.13.7/bin/ninja -C /home/warren/repos/git/forks/scipy-fresh/build -j 12
ninja: Entering directory `/home/warren/repos/git/forks/scipy-fresh/build'
[2/338] Generating subprojects/highs/src/HConfig.h with a custom command
$ meson install --only-changed -C build --destdir ../build-install --tags=runtime,python-runtime,tests,devel
$ export PYTHONPATH="/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages"
$ /home/warren/py3.13.7/bin/python3.13 -P -c 'import scipy'
$ cd /home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages
$ /home/warren/py3.13.7/bin/python3.13 -P -m pytest -n 3 --pyargs scipy.cluster
================================== test session starts ===================================
platform linux -- Python 3.13.7, pytest-8.4.2, pluggy-1.6.0
rootdir: /home/warren/repos/git/forks/scipy-fresh
configfile: pytest.ini
plugins: xdist-3.8.0, anyio-4.11.0, timeout-2.4.0, hypothesis-6.138.14, cov-6.2.1
3 workers [217 items]   
.................................................................................. [ 37%]
.................................................................................. [ 75%]
...................................................F.                              [100%]
======================================== FAILURES ========================================
________________________ TestKMeans.test_kmeans2_high_dim[numpy] _________________________
[gw0] linux -- Python 3.13.7 /home/warren/py3.13.7/bin/python3.13

self = <scipy.cluster.tests.test_vq.TestKMeans object at 0x765c415ada50>
xp = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

    def test_kmeans2_high_dim(self, xp):
        # test kmeans2 when the number of dimensions exceeds the number
        # of input points
        data = xp.asarray(TESTDATA_2D)
        data = xp.reshape(data, (20, 20))[:10, :]
>       kmeans2(data, 2)

data       = array([[-2.2 ,  1.17, -1.63,  1.69, -2.04,  4.38, -3.09,  0.95, -1.7 ,
         4.79, -1.68,  0.68, -2.26,  3.34, -2.2...75, -2.25,  1.71, -3.28,
         3.38, -1.74,  0.88, -2.41,  1.92, -2.24,  1.19, -2.48,  1.06,
        -1.68, -0.62]])
self       = <scipy.cluster.tests.test_vq.TestKMeans object at 0x765c415ada50>
xp         = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>

scipy/cluster/tests/test_vq.py:331: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
scipy/_lib/_util.py:346: in wrapper
    return fun(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^
        NEW_NAME   = 'rng'
        args       = (array([[-2.2 ,  1.17, -1.63,  1.69, -2.04,  4.38, -3.09,  0.95, -1.7 ,
         4.79, -1.68,  0.68, -2.26,  3.34, -2....-2.25,  1.71, -3.28,
         3.38, -1.74,  0.88, -2.41,  1.92, -2.24,  1.19, -2.48,  1.06,
        -1.68, -0.62]]), 2)
        as_new_kwarg = False
        as_old_kwarg = False
        as_pos_arg = False
        cmn_msg    = 'To silence this warning and ensure consistent behavior in SciPy None, control the RNG using argument `rng`. Arguments...ndom` or `RandomState` instances will result in an error. See the documentation of `default_rng` for more information.'
        emit_warning = False
        end_version = None
        fun        = <function kmeans2 at 0x765c75362de0>
        global_seed_set = False
        kwargs     = {}
        old_name   = 'seed'
        position_num = None
scipy/cluster/vq.py:827: in kmeans2
    miss_meth()
        _          = 0
        check_finite = True
        code_book  = array([[-1.94263604,  2.39670081, -3.168438  ,  2.54556934, -3.22427165,
         3.17363025, -1.81385259,  0.91074636...9365, -2.65092999, -4.20645273, -4.122284  ,
        -0.52629789, -3.57167399,  3.34294679, -0.21820847,  1.19803037]])
        d          = 20
        data       = array([[-2.2 ,  1.17, -1.63,  1.69, -2.04,  4.38, -3.09,  0.95, -1.7 ,
         4.79, -1.68,  0.68, -2.26,  3.34, -2.2...75, -2.25,  1.71, -3.28,
         3.38, -1.74,  0.88, -2.41,  1.92, -2.24,  1.19, -2.48,  1.06,
        -1.68, -0.62]])
        has_members = array([ True, False])
        init_meth  = <function _krandinit at 0x765c7ddcf920>
        iter       = 10
        k          = 2
        label      = array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int32)
        minit      = 'random'
        miss_meth  = <function _missing_warn at 0x765c75362b60>
        missing    = 'warn'
        nc         = 2
        new_code_book = array([[-2.167,  1.365, -2.194,  1.835, -2.763,  2.068, -2.037,  1.657,
        -2.779,  3.082, -2.813,  2.507, -2.237....   ,
         0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,  0.   ,
         0.   ,  0.   ,  0.   ,  0.   ]])
        rng        = RandomState(MT19937) at 0x765CCB7B8740
        thresh     = 1e-05
        xp         = <module 'scipy._lib.array_api_compat.numpy' from '/home/warren/repos/git/forks/scipy-fresh/build-install/usr/lib/python3.13/site-packages/scipy/_lib/array_api_compat/numpy/__init__.py'>
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    def _missing_warn():
        """Print a warning when called."""
>       warnings.warn("One of the clusters is empty. "
                      "Re-run kmeans with a different initialization.",
                      stacklevel=3)
E       UserWarning: One of the clusters is empty. Re-run kmeans with a different initialization.


scipy/cluster/vq.py:634: UserWarning
================================ short test summary info =================================
FAILED scipy/cluster/tests/test_vq.py::TestKMeans::test_kmeans2_high_dim[numpy] - UserWarning: One of the clusters is empty. Re-run kmeans with a different initializat...
============================= 1 failed, 216 passed in 1.68s ==============================
```

</details>

<details>

<summary> System details </summary>

```
In [1]: import sys, scipy, numpy; print(scipy.__version__, numpy.__version__, sys.version_info); scipy.sh
      ⋮ ow_config()
1.17.0.dev0+git20251115.45d05f3 2.4.0.dev0+git20250926.087615c sys.version_info(major=3, minor=13, micro=7, releaselevel='final', serial=0)
Build Dependencies:
  blas:
    detection method: pkgconfig
    found: true
    has ilp64: false
    include directory: /usr/include/x86_64-linux-gnu/openblas-pthread/
    lib directory: /usr/lib/x86_64-linux-gnu/openblas-pthread/
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS=
      NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    version: 0.3.20
  lapack:
    detection method: pkgconfig
    found: true
    has ilp64: false
    include directory: /usr/include/x86_64-linux-gnu/openblas-pthread/
    lib directory: /usr/lib/x86_64-linux-gnu/openblas-pthread/
    name: openblas
    openblas configuration: USE_64BITINT= DYNAMIC_ARCH=1 DYNAMIC_OLDER=1 NO_CBLAS=
      NO_LAPACK= NO_LAPACKE=1 NO_AFFINITY=1 USE_OPENMP=0 generic MAX_THREADS=64
    pc file directory: /usr/lib/x86_64-linux-gnu/pkgconfig
    version: 0.3.20
  pybind11:
    detection method: config-tool
    include directory: unknown
    name: pybind11
    version: 3.0.1
Compilers:
  c:
    commands: cc
    linker: ld.bfd
    name: gcc
    version: 11.4.0
  c++:
    commands: c++
    linker: ld.bfd
    name: gcc
    version: 11.4.0
  cython:
    commands: cython
    linker: cython
    name: cython
    version: 3.1.3
  fortran:
    commands: gfortran
    linker: ld.bfd
    name: gcc
    version: 11.4.0
  pythran:
    include directory: ../../../../../py3.13.7/lib/python3.13/site-packages/pythran
    version: 0.18.0
Machine Information:
  build:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
  cross-compiled: false
  host:
    cpu: x86_64
    endian: little
    family: x86_64
    system: linux
Python Information:
  path: /home/warren/py3.13.7/bin/python3.13
  version: '3.13'
```

</details>

--- a/scipy/cluster/vq/tests/test_vq.py
+++ b/scipy/cluster/vq/tests/test_vq.py
@@ -320,14 +320,16 @@ class TestKMeans:
     def test_kmeans2_rank1_2(self, xp):
         data = xp.asarray(TESTDATA_2D)
         data1 = data[:, 0]
-        kmeans2(data1, 2, iter=1)
+        rng = np.random.default_rng(42)
+        kmeans2(data1, 2, iter=1, rng=rng)
 
     def test_kmeans2_high_dim(self, xp):
         # test kmeans2 when the number of dimensions exceeds the number
         # of input points
         data = xp.asarray(TESTDATA_2D)
         data = xp.reshape(data, (20, 20))[:10, :]
-        kmeans2(data, 2)
+        rng = np.random.default_rng(42)
+        kmeans2(data, 2, rng=rng)     
 
     def test_kmeans2_init(self, xp):
         rng = np.random.default_rng(12345678)

